### PR TITLE
fixes async operations plus tag support

### PR DESCRIPTION
**Pull request by egbertn**
Bonjour Joeaudette!
Hope you find this usefull. The current 3.0.0 cannot work unless a global mvc setting allow synchroneous IO is set to true.

### CODE REVIEW

File: src/cloudscribe.MetaWeblog/Controllers/MetaWeblogController.cs [hunk 1]
```diff
@@ -71,33 +71,34 @@ public virtual async Task<IActionResult> Index()
...
             XDocument resultXml;
             MetaWeblogResult outCome;
-            FaultStruct faultStruct;
+
             try
             {
                 using (HttpContext.Request.Body)
                 {
-                    postedXml = XDocument.Load(HttpContext.Request.Body);
+                    postedXml = await XDocument.LoadAsync(HttpContext.Request.Body, LoadOptions.None, default);
```
Comment (egbertn) [inline src/cloudscribe.MetaWeblog/Controllers/MetaWeblogController.cs:80]: Load will fail. Async fails too. Need to set AllowAsynchroneousIO = true

File: src/cloudscribe.MetaWeblog/Controllers/MetaWeblogController.cs [hunk 2]
```diff
@@ -111,10 +112,11 @@ public virtual async Task<IActionResult> Index()
                     + Utils.GetDateTimeStringForFileName(true)  + "-"
                     + metaWeblogRequest.MethodName.Replace(".","-")
                     + ".xml";
-                using (StreamWriter s = System.IO.File.CreateText(requestFileName))
-                {
-                    postedXml.Save(s, SaveOptions.None);
-                }
+                using var s = System.IO.File.Create(requestFileName);
+                //await postedXml.SaveAsync(s, SaveOptions.None, default);
```
Comment (egbertn) [inline src/cloudscribe.MetaWeblog/Controllers/MetaWeblogController.cs:118]: required on .net 3.0

File: src/cloudscribe.MetaWeblog/Controllers/MetaWeblogController.cs [hunk 3]
```diff
@@ -154,10 +156,11 @@ public virtual async Task<IActionResult> Index()
...
                     + ".xml";
 
-                using (StreamWriter s = System.IO.File.CreateText(reseponseFileName))
-                {
-                    resultXml.Save(s, SaveOptions.None);
-                }
+                using var s = System.IO.File.Create(reseponseFileName);
+                // await resultXml.SaveAsync(s, SaveOptions.None, default);
+                var xmlWriter = XmlWriter.Create(s, settings: new XmlWriterSettings { Async = true });
+                await resultXml.WriteToAsync(xmlWriter, default);
```
Comment (egbertn) [inline src/cloudscribe.MetaWeblog/Controllers/MetaWeblogController.cs:164]: async required for 3.0.0

File: src/cloudscribe.MetaWeblog/IMetaWeblogService.cs [hunk 4]
```diff
@@ -51,7 +51,12 @@ string authorDisplayName
             string userName,
             string password,
             CancellationToken cancellationToken);
-        
+
+        Task<IEnumerable<TagStruct>> GetTags(
+            string blogId,
```
Comment (egbertn) [inline src/cloudscribe.MetaWeblog/IMetaWeblogService.cs]: new feature was not yet implemented. Now it is.

File: src/cloudscribe.MetaWeblog/Utils.cs [hunk 5]
```diff
@@ -1,85 +1,28 @@
...
 using System.Globalization;
-using System.Linq;
-using System.Threading.Tasks;
 
 namespace cloudscribe.MetaWeblog
 {
     public static class Utils
     {
-        public static string GetDateTimeStringForFileName()
+        public static string GetDateTimeStringForFileName(bool includeMiliseconds = false)
```
Comment (egbertn) [inline src/cloudscribe.MetaWeblog/Utils.cs:9]: could not leave this, I know, bad conduct. But I Tested output.  Its ok.

File: src/cloudscribe.MetaWeblog/XmlResult.cs [hunk 6]
```diff
@@ -32,26 +34,28 @@ public override void ExecuteResult(ActionContext context)
...
         {
             context.HttpContext.Response.ContentType = this.ContentType;
 
             if (Xml != null)
             {
-                Xml.Save(context.HttpContext.Response.Body, SaveOptions.DisableFormatting);
-                return Task.FromResult(0);
+               var xmlWriter = XmlWriter.Create(context.HttpContext.Response.Body, settings: new XmlWriterSettings { Async = true });
+
+                await Xml.WriteToAsync(xmlWriter, default);
```
Comment (egbertn) [inline src/cloudscribe.MetaWeblog/XmlResult.cs:52]: required for .net core 3.x

File: src/cloudscribe.MetaWeblog/cloudscribe.MetaWeblog.csproj [hunk 7]
```diff
@@ -11,24 +11,18 @@
...
   </ItemGroup>
 
   <ItemGroup>
     <FrameworkReference Include="Microsoft.AspNetCore.App" />
   </ItemGroup>
 
-  <ItemGroup>
-    <PackageReference Include="Microsoft.Extensions.Options" Version="3.0.0" />
-    <PackageReference Include="Microsoft.Extensions.Options.ConfigurationExtensions" Version="3.0.0" />
-    <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="3.0.0" />
```
Comment (egbertn) [inline src/cloudscribe.MetaWeblog/cloudscribe.MetaWeblog.csproj:28]: these references are not required anymore

**Comment by joeaudette** [comment]
Bonjour @egbertn !

In order to consider accepting your pull request, we need a signed contributor agreement.
https://raw.githubusercontent.com/cloudscribe/cloudscribe/master/copyright-assignment.txt

Thank you,

Joe

**Comment by egbertn** [comment]
Hi Joe,

???
If it is on github and having apache 2.0 license, why this mistrust?
If I contribute this cannot in any way lead to claims on a public repo.
Just ignore the pull request. Ok?

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

**Comment by joeaudette** [comment]
Hi Egbert,

It is not a matter of mistrust, it is a matter of keeping copyright ownership purity so that we keep our options open. We may want to give the copyrights to the .net foundation in the future, or we may want to change the license to MIT or we may want to dual license to offer our customers a commercial license. In order to do any of those things we need to continue to own the code 100%.

See also our contributing guide https://github.com/cloudscribe/cloudscribe.MetaWeblog/blob/master/CONTRIBUTING.md
https://github.com/cloudscribe/cloudscribe/blob/master/CONTRIBUTING.md

It would be the same thing if you make a pull request on any well run open source project, if you want to make a pull request to asp.net core you have to sign an agreement with Microsoft, we are not big like Microsoft but it is still just as important and valid for us to maintain full copyright ownership of our projects. Same thing if you want to contribute to any major open source project like MySql or the linux kernel, you have to sign an agreement to do that. It is not all just handled by github.

Best Regards,

Joe

**Comment by joeaudette** [comment]
FYI, I've moved the code from this repo into the cloudscribe.SimpleContent repo and fixed the async IO issues. Going forward any new work on cloudscribe.Metaweblog will happen in the SimpleContent repo
https://github.com/cloudscribe/cloudscribe.SimpleContent

The reason for this is because that repo makes it much easier to debug the metaweblog code by using SimpleContent and Open Live Writer.

Eventually we will delete the code from this repo.

@egbertn if you change your mind and decide to sign the agreement, then you can make a pull request in the SimpleContent repo to add the tags support. Otherwise you are free to fork the code and use a forked version.

**Comment by joeaudette** [comment]
I've also published an updated nuget with the IO fix https://www.nuget.org/packages/cloudscribe.MetaWeblog/3.0.2

**Comment by egbertn** [comment]
Ok will do later.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

________________________________

**Pull request closed without merging**